toolchain/language_server/context.cpp

```diff
@@ -155,6 +155,55 @@ auto Context::File::SetText(Context& context, std::optional<int64_t> version,
                             consumer.params());
 }
 
+auto Context::File::PositionToIndex(const clang::clangd::Position& position)
+    -> size_t {
+  return PositionToIndex(source_->text().data(), position);
+}
+
+auto Context::File::PositionToIndex(const std::string& contents,
+                                    const clang::clangd::Position& position)
+    -> size_t {
+  size_t result = 0;
+
+  for (int row = 0; row < position.line; row++) {
+    const size_t newline_index = contents.find('\n', result);
+
+    CARBON_CHECK(newline_index != std::string::npos,
+                 "Line number greater than number of lines in the file");
+
+    result = newline_index + 1;
+  }
+
+  result += position.character;
+
+  CARBON_CHECK(result <= contents.size(),
+               "Position greater than source code size");
+
+  return result;
+}
+
+auto Context::File::ApplyChanges(
+    Context& context, std::optional<int64_t> version,
+    const std::vector<clang::clangd::TextDocumentContentChangeEvent>&
+        content_changes) -> void {
+  std::string contents = source_->text().data();
+
+  for (const auto& change : content_changes) {
+    // if range is not present, then we copy entire text
```

Comment:
Per [punctuation style](https://google.github.io/styleguide/cppguide.html#Punctuation,_Spelling_and_Grammar), please use sentence structure in comments.

```suggestion
    // If range is not present, then we copy entire text.
```